src/Processors/Formats/Impl/Parquet/ParquetLeafColReader.cpp

```diff
@@ -0,0 +1,521 @@
+    {
+        // if dictionary page encountered, another page should be read
+        readPageIfEmpty();
+
+        auto read_values = static_cast<UInt32>(std::min(rows_num, static_cast<UInt64>(cur_page_values)));
+        data_values_reader->readBatch(column, *null_map, read_values);
+
+        cur_page_values -= read_values;
+        rows_num -= read_values;
+    }
+
+    return releaseColumn(name);
+}
+
+template <>
+void ParquetLeafColReader<ColumnString>::resetColumn(UInt64 rows_num)
+{
+    if (reading_low_cardinality)
+    {
+        assert(dictionary);
+        visitColStrIndexType(dictionary->size(), [&]<typename TColVec>(TColVec *)
+        {
+            column = TColVec::create();
+        });
+
+        // only first position is used
+        null_map = std::make_unique<LazyNullMap>(1);
+        column->reserve(rows_num);
+    }
+    else
+    {
+        null_map = std::make_unique<LazyNullMap>(rows_num);
+        column = ColumnString::create();
+        reserveColumnStrRows(column, rows_num);
+    }
+}
+
+template <typename TColumn>
+void ParquetLeafColReader<TColumn>::resetColumn(UInt64 rows_num)
+{
+    assert(!reading_low_cardinality);
+
+    column = base_data_type->createColumn();
+    column->reserve(rows_num);
+    null_map = std::make_unique<LazyNullMap>(rows_num);
+}
+
+template <typename TColumn>
+void ParquetLeafColReader<TColumn>::degradeDictionary()
+{
+    // if last batch read all dictionary indices, then degrade is not needed this time
+    if (!column)
+    {
+        dictionary = nullptr;
+        return;
+    }
+    assert(dictionary && !column->empty());
+
+    null_map = std::make_unique<LazyNullMap>(reading_rows_num);
+    auto col_existing = std::move(column);
+    column = ColumnString::create();
+
+    ColumnString & col_dest = *static_cast<ColumnString *>(column.get());
+    const ColumnString & col_dict_str = *static_cast<const ColumnString *>(dictionary.get());
+
+    visitColStrIndexType(dictionary->size(), [&]<typename TColVec>(TColVec *)
+    {
+        const TColVec & col_src = *static_cast<const TColVec *>(col_existing.get());
+        reserveColumnStrRows(column, reading_rows_num);
+
+        col_dest.getOffsets().resize(col_src.size());
+        for (size_t i = 0; i < col_src.size(); i++)
+        {
+            auto src_idx = col_src.getData()[i];
+            if (0 == src_idx)
+            {
+                null_map->setNull(i);
+            }
+            auto dict_chars_cursor = col_dict_str.getOffsets()[src_idx - 1];
+            auto str_len = col_dict_str.getOffsets()[src_idx] - dict_chars_cursor;
+            auto dst_chars_cursor = col_dest.getChars().size();
+            col_dest.getChars().resize(dst_chars_cursor + str_len);
+
+            memcpySmallAllowReadWriteOverflow15(
+                &col_dest.getChars()[dst_chars_cursor], &col_dict_str.getChars()[dict_chars_cursor], str_len);
+            col_dest.getOffsets()[i] = col_dest.getChars().size();
+        }
+    });
```

Comment:
Well, it will be easier. But the current implementation is more efficient, even not very much. Because the size of new Column generated by `convertToFullColumn` is smaller than `reading_rows_num`, so it should be resized to `reading_rows_num` again.

By the way, I will implement the logic as your suggestion next time. For current codes, I'd prefer not to change it right now. Instead, I can add some comments.